extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -272,153 +337,142 @@ public void run()
+                    record.offset(),
+                    nextOffsets.get(record.partition()),
+                    record.partition()
+                );
               }
 
-              fireDepartmentMetrics.incrementProcessed();
-            }
-            catch (ParseException e) {
-              if (tuningConfig.isReportParseExceptions()) {
-                throw e;
-              } else {
-                log.debug(
-                    e,
-                    "Dropping unparseable row from partition[%d] offset[%,d].",
-                    record.partition(),
-                    record.offset()
+              try {
+                final InputRow row = Preconditions.checkNotNull(parser.parse(ByteBuffer.wrap(record.value())), "row");
+                final SegmentIdentifier identifier = driver.add(
+                    row,
+                    sequenceNames.get(record.partition()),
+                    committerSupplier
                 );
 
-                fireDepartmentMetrics.incrementUnparseable();
-              }
-            }
+                if (identifier == null) {
+                  // Failure to allocate segment puts determinism at risk, bail out to be safe.
+                  // May want configurable behavior here at some point.
+                  // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
+                  throw new ISE("Could not allocate segment for row with timestamp[%s]", row.getTimestamp());
+                }
 
-            final long nextOffset = record.offset() + 1;
-            final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(record.partition());
+                fireDepartmentMetrics.incrementProcessed();
+              }
+              catch (ParseException e) {
+                if (tuningConfig.isReportParseExceptions()) {
+                  throw e;
+                } else {
+                  log.debug(
+                      e,
+                      "Dropping unparseable row from partition[%d] offset[%,d].",
+                      record.partition(),
+                      record.offset()
+                  );
+
+                  fireDepartmentMetrics.incrementUnparseable();
+                }
+              }
 
-            nextOffsets.put(record.partition(), nextOffset);
+              nextOffsets.put(record.partition(), record.offset() + 1);
+            }
 
-            if (nextOffset == endOffset && assignment.remove(record.partition())) {
+            if (nextOffsets.get(record.partition()).equals(endOffsets.get(record.partition()))
+                && assignment.remove(record.partition())) {
               log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
               assignPartitions(consumer, topic, assignment);
-              stillReading = !assignment.isEmpty();
+              stillReading = ioConfig.isPauseAfterRead() || !assignment.isEmpty();
```

Comment:
If `pauseAfterRead` is true, `stillReading` will remain true regardless of the state of `assignment` - this will ensure we continue to loop and will pause in `possiblyPause()` if all partitions have been read. If `pauseAfterRead` is false, the loop will end as soon as all the partitions have been read.